framework/mongo/handlers.py

```diff
@@ -33,8 +32,11 @@ def acquire(self, _id=None):
         return self._local[_id]
 
     def release(self, _id=None):
-        self._cache.append(self._local.pop(_id or self.thread_id))
-        self._sem.release()
+        _id = _id or self.thread_id
+
+        if _id in self._local:
```

Comment:
Release should not be called if this line is false. It should raise an error if you're attempting to release a connection you don't own